File: backend/calculations.py
"""
All financial metric computations.
Each function is pure (no side-effects) and documented with its formula.
"""
from __future__ import annotations

import logging
from datetime import date, datetime
from typing import Optional

import numpy as np
import pandas as pd
from pyxirr import xirr

from config import Settings
from excel_loader import INCOME_TYPES, WorkbookData
from models import (
    Allocation,
    AllocationSlice,
    DataWarning,
    DistributionEvent,
    Distributions,
    Holdings,
    Holding,
    IRR,
    KPIs,
    NavPoint,
    RiskMetrics,
    TargetVsActual,
    TimeSeries,
)

logger = logging.getLogger(__name__)


# ── Helpers ───────────────────────────────────────────────────────────────────

def _to_date(val) -> date:
    if isinstance(val, date):
        return val
    return pd.Timestamp(val).date()


def _nav_series(data: WorkbookData) -> pd.Series:
    """Return Portfolio Value indexed by Month End timestamp."""
    mr = data.monthly_returns.set_index("Month End")["Portfolio Value"]
    return mr.sort_index()


# ── KPIs ──────────────────────────────────────────────────────────────────────

def compute_kpis(data: WorkbookData, settings: Settings) -> KPIs:
    pm = data.portfolio_metrics
    tl = data.trade_log

    # Total Portfolio Value = Holdings market value + cash (pre-computed)
    total_portfolio_value: float = float(pm.get("Total Portfolio Value", 0))

    # Capital committed = total deposits ever made
    capital_committed: float = float(pm.get("Total Invested Capital", 0))

    # % since entry = (NAV - committed) / committed
    pct_since_entry: float = (total_portfolio_value - capital_committed) / capital_committed

    # MOIC = (NAV + total income received) / capital_committed
    # Formula: total return to investor if they liquidated today (NAV) plus
    # cash already received (income), divided by total cash ever put in.
    total_income: float = float(pm.get("Total Income (Div+Cpn+Dist)", 0))
    moic: float = (total_portfolio_value + total_income) / capital_committed

    # Current yield = TTM income / NAV
    # We use total income as a proxy for TTM (portfolio is ~15 months old)
    current_yield: float = total_income / total_portfolio_value if total_portfolio_value else 0.0

    # Distributions (all income types) from trade log
    income_rows = tl[tl["Type"].isin(INCOME_TYPES)].copy()
    # Net out negative rows (withholding tax rows paired with gross distribution)
    distributions_total: float = float(income_rows["Net Amount"].sum())
    distributions_count: int = len(income_rows[income_rows["Net Amount"] > 0])
    last_income = income_rows[income_rows["Net Amount"] > 0]["Date"]
    distributions_last_date: Optional[date] = (
        _to_date(last_income.max()) if not last_income.empty else None
    )

    return KPIs(
        total_portfolio_value=total_portfolio_value,
        capital_committed=capital_committed,
        pct_since_entry=pct_since_entry,
        moic=moic,
        moic_target=settings.moic_target,
        current_yield=current_yield,
        distributions_total=distributions_total,
        distributions_count=distributions_count,
        distributions_last_date=distributions_last_date,
    )


# ── Time series ───────────────────────────────────────────────────────────────

def compute_timeseries(data: WorkbookData) -> TimeSeries:
    """
    Monthly NAV series from Monthly Returns sheet.
    Normalized index: nav[t] / nav[0] * 100, so the first month = 100.
    """
    mr = data.monthly_returns.copy()
    if mr.empty:
        return TimeSeries(series=[])

    nav0 = mr.iloc[0]["Portfolio Value"]
    points: list[NavPoint] = []
    for _, row in mr.iterrows():
        nav = float(row["Portfolio Value"])
        normalized = (nav / nav0 * 100) if nav0 else 100.0
        points.append(NavPoint(
            month_end=_to_date(row["Month End"]),
            nav=nav,
            normalized=round(normalized, 4),
            monthly_return=float(row["Monthly Return"]),
            cumulative_return=float(row["Cumulative Return"]),
        ))
    return TimeSeries(series=points)


# ── Allocation ────────────────────────────────────────────────────────────────

# Maps Holdings "Asset Class" values → canonical dashboard labels
_CLASS_LABEL: dict[str, str] = {
    "Stock": "Stocks",
    "Bond": "Bonds",
    "ETF/ETC": "ETFs / ETCs",
    "Crypto ETP": "Crypto ETPs",
}


def compute_allocation(data: WorkbookData) -> Allocation:
    """
    Group Holdings by Asset Class, sum Market Value, compute weight.
    Add Cash from Portfolio Metrics (not represented as a holding).
    """
    pm = data.portfolio_metrics
    h = data.holdings[data.holdings["Market Value"] > 0].copy()

    groups = (
        h.groupby("Asset Class")["Market Value"]
        .sum()
        .rename(index=_CLASS_LABEL)
        .to_dict()
    )
    cash = float(pm.get("Cash Balance (est.)", 0))
    if cash > 0:
        groups["Cash"] = cash

    total = sum(groups.values())
    slices = [
        AllocationSlice(
            asset_class=ac,
            market_value=round(mv, 2),
            weight=round(mv / total, 6) if total else 0.0,
        )
        for ac, mv in sorted(groups.items(), key=lambda x: -x[1])
    ]
    return Allocation(slices=slices)


# ── IRR ───────────────────────────────────────────────────────────────────────

def _xirr_safe(dates: list[date], cashflows: list[float]) -> Optional[float]:
    """
    Wrapper around pyxirr.xirr.
    Returns None if convergence fails or inputs are degenerate.
    pyxirr expects dates as date objects and amounts as floats.
    Sign convention: outflows negative, inflows positive.
    """
    try:
        result = xirr(dates, cashflows)
        if result is None or not np.isfinite(result):
            return None
        return float(result)
    except Exception as exc:
        logger.warning("XIRR failed: %s", exc)
        return None


def _build_investor_cashflows(data: WorkbookData) -> tuple[list[date], list[float]]:
    """
    Investor perspective cash flows:
      - Deposits as outflows (negative from investor's wallet)
      - Withdrawals as inflows (positive back to investor)
      - Terminal value = Total Portfolio Value (NAV + cash) as final inflow
    """
    inv = data.irr_investor.copy()
    dates = [_to_date(d) for d in inv["Date"]]
    flows = [float(cf) for cf in inv["Cash Flow"]]

    # Append terminal value at cutoff date
    nav = float(data.portfolio_metrics.get("Total Portfolio Value", 0))
    dates.append(_to_date(data.cutoff_date))
    flows.append(nav)
    return dates, flows


def _build_portfolio_cashflows(data: WorkbookData) -> tuple[list[date], list[float]]:
    """
    Portfolio (fund) perspective cash flows:
      - Buy/investment outflows (negative)
      - Sell/income inflows (positive)
      - Terminal value = Market Value of open positions (no cash) as final inflow
    """
    port = data.irr_portfolio.copy()
    dates = [_to_date(d) for d in port["Date"]]
    flows = [float(cf) for cf in port["Cash Flow"]]

    mktval = float(data.portfolio_metrics.get("Total Market Value (Holdings)", 0))
    dates.append(_to_date(data.cutoff_date))
    flows.append(mktval)
    return dates, flows


def compute_irr(data: WorkbookData) -> IRR:
    """
    Compute XIRR for both investor and fund perspectives using pyxirr.
    Falls back to pre-computed Excel values if XIRR fails to converge.
    """
    pm = data.portfolio_metrics

    inv_dates, inv_flows = _build_investor_cashflows(data)
    investor_irr = _xirr_safe(inv_dates, inv_flows)
    if investor_irr is None:
        logger.warning("Investor XIRR failed — using pre-computed value from Excel.")
        investor_irr = float(pm.get("Investor IRR", 0))

    port_dates, port_flows = _build_portfolio_cashflows(data)
    fund_irr = _xirr_safe(port_dates, port_flows)
    if fund_irr is None:
        logger.warning("Fund XIRR failed — using pre-computed value from Excel.")
        fund_irr = float(pm.get("Company IRR", 0))

    return IRR(
        fund_irr=round(fund_irr, 6),
        investor_irr=round(investor_irr, 6),
        valuation_date=_to_date(data.cutoff_date),
    )


# ── Risk metrics ──────────────────────────────────────────────────────────────

def compute_risk(data: WorkbookData, settings: Settings) -> RiskMetrics:
    """
    Sharpe  = (annualized_return - risk_free_rate) / annualized_volatility
    Volatility (ann) = std(monthly_returns, ddof=1) * sqrt(12)

    Max drawdown (CFA/GIPS standard for portfolios with external cash flows):
      Build a TWR index from the monthly return series, then compute the
      peak-to-trough decline on that index.
      Using raw NAV would overstate peaks (deposits inflate NAV but are not
      performance), so TWR index is the correct basis.
      TWR_index[0] = 1.0; TWR_index[t] = TWR_index[t-1] * (1 + r_t)
      MDD = min( TWR_index[t] / running_max(TWR_index[:t]) - 1 )

    Risk-free rate: read from Excel Portfolio Metrics sheet first; fall back
    to config value if the cell is absent (e.g. on an older workbook version).
    """
    mr = data.monthly_returns.copy()
    returns = mr["Monthly Return"].dropna().values

    monthly_std = float(np.std(returns, ddof=1)) if len(returns) > 1 else 0.0
    vol_ann = monthly_std * np.sqrt(12)

    ann_return = float(data.portfolio_metrics.get("Annualized Return (TWR)", 0))

    # Risk-free rate: Excel is authoritative; config is the fallback
    rf_from_excel = data.portfolio_metrics.get("Risk-Free Rate (annual)")
    rf = float(rf_from_excel) if rf_from_excel is not None else settings.risk_free_rate

    sharpe = (ann_return - rf) / vol_ann if vol_ann > 1e-10 else 0.0

    # TWR index: compound the monthly returns (first period return may be 0)
    twr_index = np.cumprod(1 + returns)
    running_max = np.maximum.accumulate(twr_index)
    drawdowns = twr_index / running_max - 1
    max_drawdown = float(np.min(drawdowns)) if len(drawdowns) > 0 else 0.0

    return RiskMetrics(
        sharpe_ratio=round(sharpe, 5),
        volatility_annualized=round(vol_ann, 6),
        max_drawdown=round(max_drawdown, 6),
        annualized_return=round(ann_return, 6),
        risk_free_rate=rf,
        data_window_months=len(mr),
        beta_vs_msci_world=None,   # no benchmark data in workbook
    )


# ── Distributions ─────────────────────────────────────────────────────────────

def compute_distributions(data: WorkbookData) -> Distributions:
    """
    All income events from Trade Log (Dividend, Coupon, Distribution, Sec. Lending).
    Rows with negative Net Amount are withholding-tax offsets — included to give
    accurate net totals; filtered from the event list if amount ≤ 0.
    """
    tl = data.trade_log
    income = tl[tl["Type"].isin(INCOME_TYPES)].copy()
    income = income.sort_values("Date")

    events = [
        DistributionEvent(
            date=_to_date(row["Date"]),
            security=str(row["Security"]),
            income_type=str(row["Type"]),
            amount=round(float(row["Net Amount"]), 2),
        )
        for _, row in income.iterrows()
    ]
    total = round(float(income["Net Amount"].sum()), 2)
    return Distributions(events=events, total=total)


# ── Target vs Actual ──────────────────────────────────────────────────────────

def compute_targets(data: WorkbookData, settings: Settings) -> TargetVsActual:
    """
    Actual weights come from Portfolio Metrics asset-class block.
    Alternatives = ETF/ETC + Crypto ETP (combined to match the 3-bucket target).
    """
    pm = data.portfolio_metrics
    total_nav = float(pm.get("Total Portfolio Value", 1))

    def _w(label: str) -> float:
        v = pm.get(label, 0)
        return float(v) / total_nav if total_nav else 0.0

    # Asset class market values from the pre-computed block
    h = data.holdings
    stock_mv = float(h[h["Asset Class"] == "Stock"]["Market Value"].sum())
    bond_mv = float(h[h["Asset Class"] == "Bond"]["Market Value"].sum())
    etf_mv = float(h[h["Asset Class"] == "ETF/ETC"]["Market Value"].sum())
    crypto_mv = float(h[h["Asset Class"] == "Crypto ETP"]["Market Value"].sum())
    cash_mv = float(pm.get("Cash Balance (est.)", 0))

    equity_pct = stock_mv / total_nav
    bond_pct = bond_mv / total_nav
    alt_pct = (etf_mv + crypto_mv) / total_nav
    cash_pct = cash_mv / total_nav

    return TargetVsActual(
        target_equity_pct=settings.target_equity_pct,
        target_bond_pct=settings.target_bond_pct,
        target_alt_pct=settings.target_alt_pct,
        current_equity_pct=round(equity_pct, 6),
        current_bond_pct=round(bond_pct, 6),
        current_alt_pct=round(alt_pct, 6),
        current_cash_pct=round(cash_pct, 6),
    )


# ── Holdings ──────────────────────────────────────────────────────────────────

def compute_holdings(data: WorkbookData) -> Holdings:
    h = data.holdings.copy()
    # Include all rows (including zero-share closed positions) for audit purposes
    items = [
        Holding(
            security=str(row["Security"]),
            asset_class=str(row["Asset Class"]),
            currency=str(row["Currency"]),
            shares=float(row["Shares Held"]),
            avg_cost=float(row["Avg Cost"]),
            cost_basis=float(row["Total Cost Basis"]),
            current_price=float(row["Current Price"]) if row["Current Price"] != 0 else None,
            market_value=float(row["Market Value"]),
            unrealized_pnl=float(row["Unrealized P&L"]),
            pnl_pct=float(row["P&L %"]),
            weight=float(row["Weight %"]),
        )
        for _, row in h.iterrows()
    ]
    return Holdings(
        holdings=items,
        total_market_value=round(float(h["Market Value"].sum()), 3),
        total_cost_basis=round(float(h["Total Cost Basis"].sum()), 3),
        total_unrealized_pnl=round(float(h["Unrealized P&L"].sum()), 3),
        cutoff_date=_to_date(data.cutoff_date),
    )


# ── Reconciliation check ──────────────────────────────────────────────────────

def check_nav_reconciliation(data: WorkbookData, tolerance: float) -> list[DataWarning]:
    """
    Compare Holdings sum of Market Value against Portfolio Metrics Total Market Value.
    Warn if discrepancy exceeds tolerance fraction.
    """
    warnings: list[DataWarning] = []
    holdings_mv = float(data.holdings["Market Value"].sum())
    pm_mv = float(data.portfolio_metrics.get("Total Market Value (Holdings)", 0))

    if pm_mv == 0:
        return warnings

    diff_pct = abs(holdings_mv - pm_mv) / pm_mv
    if diff_pct > tolerance:
        warnings.append(DataWarning(
            code="NAV_RECONCILIATION_MISMATCH",
            message=(
                f"Holdings market value ({holdings_mv:,.2f}) differs from "
                f"Portfolio Metrics NAV ({pm_mv:,.2f}) by {diff_pct:.2%}, "
                f"exceeding {tolerance:.2%} threshold."
            ),
        ))
    return warnings


# ── Beta note ─────────────────────────────────────────────────────────────────

BETA_UNAVAILABLE_WARNING = DataWarning(
    code="BETA_UNAVAILABLE",
    message="beta_vs_msci_world is null: no MSCI World benchmark data found in the workbook.",
)

File: Investor Portal Final MVP/src/lib/calculations.ts
import type { WorkbookData, TradeRow } from "./excel-loader";
import { INCOME_TYPES } from "./excel-loader";
import type {
  KPIs,
  NavPoint,
  AllocationSlice,
  IRRData,
  RiskMetrics,
  DistributionEvent,
  TargetVsActual,
  Holding,
  PortfolioComposition,
} from "@/types/portfolio";

interface Settings {
  riskFreeRate: number;
  moicTarget: number;
  targetEquityPct: number;
  targetBondPct: number;
  targetAltPct: number;
}

// Newton-Raphson XIRR — used as fallback when pre-computed value unavailable
function xirr(cashflows: { date: Date; amount: number }[]): number | null {
  if (cashflows.length < 2) return null;
  const t0 = cashflows[0].date.getTime();
  const years = cashflows.map((cf) => (cf.date.getTime() - t0) / (365.25 * 24 * 3600 * 1000));
  const amounts = cashflows.map((cf) => cf.amount);

  function npv(rate: number): number {
    return amounts.reduce((acc, amt, i) => acc + amt / Math.pow(1 + rate, years[i]), 0);
  }
  function dnpv(rate: number): number {
    return amounts.reduce(
      (acc, amt, i) => acc - (amt * years[i]) / Math.pow(1 + rate, years[i] + 1),
      0
    );
  }

  let rate = 0.1;
  for (let iter = 0; iter < 200; iter++) {
    const f = npv(rate);
    const df = dnpv(rate);
    if (Math.abs(df) < 1e-12) break;
    const newRate = rate - f / df;
    if (Math.abs(newRate - rate) < 1e-8) return newRate;
    rate = newRate;
    if (rate < -0.9999) rate = -0.9999;
  }
  return null;
}

// Read a numeric value from portfolioMetrics by any of the given keys
function pm(data: WorkbookData, ...keys: string[]): number {
  for (const k of keys) {
    const v = data.portfolioMetrics[k];
    if (v !== undefined && v !== null && v !== "") return Number(v);
  }
  return 0;
}

function classifyAsset(assetClass: string): "equity" | "bonds" | "alts" | "cash" | "other" {
  const ac = assetClass.toLowerCase();
  if (ac.includes("stock") || ac.includes("equity")) return "equity";
  if (ac.includes("etf") || ac.includes("etc")) return "equity"; // equity ETFs grouped as equity
  if (ac.includes("bond") || ac.includes("fixed") || ac.includes("coupon")) return "bonds";
  if (ac.includes("crypto") || ac.includes("alternative") || ac.includes("etp")) return "alts";
  if (ac.includes("cash") || ac.includes("liquidity")) return "cash";
  return "other";
}

export function computeKPIs(data: WorkbookData, settings: Settings): KPIs {
  const nav = pm(data, "Total Portfolio Value", "NAV", "Portfolio Value");
  const committed = pm(data, "Capital Committed", "Total Invested Capital", "Committed Capital");

  // Total income: Div/Int Lordi Cum from tracker (already in portfolioMetrics)
  const pmIncome = pm(data, "Total Income (Div+Cpn+Dist)", "Total Income", "Total Distributions");
  const incomeRows = data.tradeLog.filter((r) => INCOME_TYPES.has(r.type));
  const tradeLogIncome = incomeRows.reduce((s, r) => s + r.netAmount, 0);
  const totalIncome = pmIncome > 0 ? pmIncome : Math.abs(tradeLogIncome);

  const pctSinceEntry = committed > 0 ? (nav - committed) / committed : 0;

  // MOIC = NAV / Capital (income is already inside NAV as cash)
  const moic = committed > 0 ? nav / committed : 0;

  const currentYield = nav > 0 ? totalIncome / nav : 0;

  const distributions = incomeRows.sort((a, b) => b.date.getTime() - a.date.getTime());
  const lastDate = distributions.length > 0 ? distributions[0].date.toISOString().split("T")[0] : null;

  return {
    totalPortfolioValue: nav,
    capitalCommitted: committed,
    pctSinceEntry,
    moic,
    moicTarget: settings.moicTarget,
    currentYield,
    distributionsTotal: totalIncome,
    distributionsCount: incomeRows.length,
    distributionsLastDate: lastDate,
    totalIncome,
  };
}

export function computeTimeseries(data: WorkbookData): NavPoint[] {
  if (data.monthlyReturns.length === 0) return [];
  const base = data.monthlyReturns[0].nav;
  let cumReturn = 0;
  return data.monthlyReturns.map((r) => {
    const normalized = base > 0 ? (r.nav / base) * 100 : 100;
    cumReturn = base > 0 ? (r.nav - base) / base : 0;
    return {
      monthEnd: r.monthEnd.toISOString().split("T")[0],
      nav: r.nav,
      normalized,
      monthlyReturn: r.monthlyReturn,
      cumulativeReturn: cumReturn,
    };
  });
}

function normalizeAssetClass(raw: string): string {
  const r = raw.toLowerCase();
  if (r.includes("stock") || r.includes("equity")) return "Stocks";
  if (r.includes("bond") || r.includes("fixed income")) return "Bonds";
  if (r.includes("etf") || r.includes("etc")) return "ETFs / ETCs";
  if (r.includes("crypto") || r.includes("etp")) return "Crypto ETPs";
  if (r.includes("cash") || r.includes("liquidity")) return "Cash";
  if (r.includes("fund") || r.includes("private") || r.includes("alternative")) return "Alternatives";
  return raw || "Other";
}

export function computeAllocation(data: WorkbookData): AllocationSlice[] {
  const buckets: Record<string, number> = {};

  for (const h of data.holdings) {
    if (h.shares <= 0) continue;
    const label = normalizeAssetClass(h.assetClass);
    buckets[label] = (buckets[label] ?? 0) + h.marketValue;
  }

  // Non-listed participations and loans as Private / Non-Listed bucket
  const nlVal = pm(data, "Non-Listed Total", "Non-Listed Value");
  const loansVal = pm(data, "Loans Outstanding");
  const privateTotal = nlVal + loansVal;
  if (privateTotal > 0) {
    buckets["Private / Non-Listed"] = (buckets["Private / Non-Listed"] ?? 0) + privateTotal;
  }

  // Cash
  const cashVal = pm(data, "Total Cash", "Cash Balance (est.)", "Cash", "Cash Total", "Liquidity");
  if (cashVal > 0) buckets["Cash"] = (buckets["Cash"] ?? 0) + cashVal;

  const total = Object.values(buckets).reduce((s, v) => s + v, 0);
  return Object.entries(buckets)
    .map(([assetClass, marketValue]) => ({
      assetClass,
      marketValue,
      weight: total > 0 ? marketValue / total : 0,
    }))
    .sort((a, b) => b.marketValue - a.marketValue);
}

export function computeIRR(data: WorkbookData): IRRData {
  const nav = pm(data, "Total Portfolio Value", "NAV", "Portfolio Value");
  const valuationDate = data.cutoffDate.toISOString().split("T")[0];

  // Prefer pre-computed IRR from 11_Performance (exact same as Excel)
  let investorIrr: number | null = null;
  const pmIrr = pm(data, "Investor IRR", "IRR Investor");
  if (pmIrr !== 0) {
    investorIrr = pmIrr > 1 ? pmIrr / 100 : pmIrr;
  }

  // Attempt XIRR from subscription cashflows (adds current NAV as terminal)
  if (investorIrr === null && data.irrInvestor.length > 1) {
    const cfs = [
      ...data.irrInvestor.map((r) => ({ date: r.date, amount: r.cashFlow })),
      { date: data.cutoffDate, amount: nav },
    ];
    investorIrr = xirr(cfs);
  }

  const fundIrr = investorIrr; // tracker uses investor perspective only

  return { fundIrr, investorIrr, valuationDate };
}

export function computeRisk(data: WorkbookData, settings: Settings): RiskMetrics {
  const returns = data.monthlyReturns.map((r) => r.monthlyReturn);
  const n = returns.length;

  // Use pre-computed risk values from 11_Performance when available
  const pmSharpe   = pm(data, "Sharpe Ratio Precomputed");
  const pmVol      = pm(data, "Volatility Annualized");
  const pmDrawdown = pm(data, "Max Drawdown Precomputed");
  const pmAnnRet   = pm(data, "Annualized Return (TWR)", "Return Total");
  const rfRate     = pm(data, "Risk Free Rate") || settings.riskFreeRate;

  if (pmVol !== 0 || pmSharpe !== 0 || pmDrawdown !== 0) {
    return {
      sharpeRatio: pmSharpe,
      volatilityAnnualized: pmVol,
      maxDrawdown: pmDrawdown,
      annualizedReturn: pmAnnRet,
      riskFreeRate: rfRate,
      dataWindowMonths: n,
    };
  }

  // Fallback: compute from monthly returns series
  if (n < 3) {
    return {
      sharpeRatio: 0,
      volatilityAnnualized: 0,
      maxDrawdown: 0,
      annualizedReturn: pmAnnRet > 1 ? pmAnnRet / 100 : pmAnnRet,
      riskFreeRate: rfRate,
      dataWindowMonths: n,
    };
  }

  const mean = returns.reduce((s, r) => s + r, 0) / n;
  const variance = returns.reduce((s, r) => s + Math.pow(r - mean, 2), 0) / (n - 1);
  const volatilityAnnualized = Math.sqrt(variance * 12);
  const ar = pmAnnRet > 1 ? pmAnnRet / 100 : pmAnnRet > 0 ? pmAnnRet : mean * 12;
  const sharpeRatio = volatilityAnnualized > 0 ? (ar - rfRate) / volatilityAnnualized : 0;

  let maxDrawdown = 0;
  let peak = data.monthlyReturns[0]?.nav ?? 0;
  for (const r of data.monthlyReturns) {
    if (r.nav > peak) peak = r.nav;
    const dd = peak > 0 ? (r.nav - peak) / peak : 0;
    if (dd < maxDrawdown) maxDrawdown = dd;
  }

  return {
    sharpeRatio,
    volatilityAnnualized,
    maxDrawdown,
    annualizedReturn: ar,
    riskFreeRate: rfRate,
    dataWindowMonths: n,
  };
}

export function computeDistributions(data: WorkbookData): DistributionEvent[] {
  return data.tradeLog
    .filter((r) => INCOME_TYPES.has(r.type))
    .map((r) => ({
      date: r.date.toISOString().split("T")[0],
      security: r.security,
      incomeType: r.type,
      amount: Math.abs(r.netAmount),
    }))
    .sort((a, b) => b.date.localeCompare(a.date));
}

export function computeTargets(data: WorkbookData, settings: Settings): TargetVsActual {
  const nav = pm(data, "Total Portfolio Value", "NAV", "Portfolio Value");
  const buckets = { equity: 0, bonds: 0, alts: 0, cash: 0 };

  for (const h of data.holdings) {
    if (h.shares <= 0) continue;
    const cls = classifyAsset(h.assetClass);
    if (cls === "equity") buckets.equity += h.marketValue;
    else if (cls === "bonds") buckets.bonds += h.marketValue;
    else if (cls === "alts") buckets.alts += h.marketValue;
  }

  // Add non-listed participations and loans to alts
  const nlVal = pm(data, "Non-Listed Total", "Non-Listed Value");
  const loansVal = pm(data, "Loans Outstanding");
  buckets.alts += nlVal + loansVal;

  buckets.cash = pm(data, "Total Cash", "Cash Balance (est.)", "Cash", "Cash Total", "Liquidity");
  const total = nav > 0 ? nav : Object.values(buckets).reduce((s, v) => s + v, 0);

  return {
    targetEquityPct: settings.targetEquityPct,
    targetBondPct: settings.targetBondPct,
    targetAltPct: settings.targetAltPct,
    currentEquityPct: total > 0 ? buckets.equity / total : 0,
    currentBondPct: total > 0 ? buckets.bonds / total : 0,
    currentAltPct: total > 0 ? buckets.alts / total : 0,
    currentCashPct: total > 0 ? buckets.cash / total : 0,
  };
}

export function computeHoldings(data: WorkbookData): Holding[] {
  return data.holdings.map((h) => ({
    security: h.security,
    assetClass: normalizeAssetClass(h.assetClass),
    currency: h.currency,
    shares: h.shares,
    avgCost: h.avgCost,
    costBasis: h.costBasis,
    currentPrice: h.currentPrice,
    marketValue: h.marketValue,
    unrealizedPnl: h.unrealizedPnl,
    pnlPct: h.pnlPct,
    weight: h.weight,
  }));
}

export function computeComposition(data: WorkbookData): PortfolioComposition {
  const listed    = pm(data, "Listed Market Value", "Listed Value");
  const nlVal     = pm(data, "Non-Listed Total", "Non-Listed Value");
  const loansVal  = pm(data, "Loans Outstanding");
  const nonListed = nlVal + loansVal;
  const cash      = pm(data, "Total Cash", "Cash Balance (est.)", "Cash", "Cash Total", "Liquidity");
  return { listed, nonListed, cash, total: listed + nonListed + cash };
}

export function checkWarnings(data: WorkbookData, tolerance = 0.01): string[] {
  const warnings: string[] = [];
  const nav = pm(data, "Total Portfolio Value", "NAV");
  const listedMv = pm(data, "Listed Market Value");
  const nonListed = pm(data, "Non-Listed Total", "Non-Listed Value");
  const loans = pm(data, "Loans Outstanding");
  const cash = pm(data, "Total Cash", "Cash Balance (est.)", "Cash");
  const reconstructed = listedMv + nonListed + loans + cash;
  if (nav > 0 && reconstructed > 0) {
    const diff = Math.abs(nav - reconstructed) / nav;
    if (diff > tolerance) {
      warnings.push(
        `NAV reconciliation: tracker NAV ${nav.toFixed(0)} vs components sum ${reconstructed.toFixed(0)} (${(diff * 100).toFixed(1)}% gap)`
      );
    }
  }
  return warnings;
}
